codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ConstrainedMapGenerator.kt

```diff
@@ -130,14 +132,22 @@ class ConstrainedMapGenerator(
             valueShape !is StructureShape &&
             valueShape !is UnionShape
         ) {
+            val keyShape = model.expectShape(shape.key.target, StringShape::class.java)
+            val keyNeedsConversion = keyShape.typeNameContainsNonPublicType(model, symbolProvider, publicConstrainedTypes)
+            val key = if (keyNeedsConversion) {
+                "k.into()"
+            } else {
+                "k"
+            }
+
```

Comment:
We had a bug here too? Maps whose keys are constrained strings but not enums need this conversion right?